posthog/models/history_logging.py

```diff
@@ -0,0 +1,191 @@
+import dataclasses
+from itertools import zip_longest
+from typing import Dict, Iterable, List, Optional, Tuple, Union
+
+from rest_framework import serializers
+
+from posthog.models import HistoricalVersion, User
+
+
+class HistoryListItemSerializer(serializers.Serializer):
+    email = serializers.EmailField(read_only=True)
+    name = serializers.CharField(read_only=True)
+    user_id = serializers.IntegerField(read_only=True)
+    action = serializers.CharField(read_only=True)
+    detail = serializers.DictField(read_only=True)
+    created_at = serializers.CharField(read_only=True)
+
+
+@dataclasses.dataclass(frozen=True)
+class HistoryListItem:
+    email: Optional[str]
+    name: Optional[str]
+    action: str
+    detail: Dict[str, Union[int, str, Dict]]
+    created_at: str
+
+
+# from https://stackoverflow.com/a/4628446/222163
+def pairwise(
+    historical_versions: List[HistoricalVersion],
+) -> Iterable[Tuple[HistoricalVersion, Union[HistoricalVersion, None]]]:
+    left = iter(historical_versions)
+    right = iter(historical_versions[1:])
+    return zip_longest(left, right)
+
+
+def load_history(history_type: str, team_id: int, item_id: int, instance, serializer):
+    """
+     * loads a page of history for that type and id (newest first)
+     * and returns a computed history for that page of history
+    """
+
+    # In order to make a page of up to 10 we need to get up to 11 as we need N-1 to determine what changed
+    versions = list(
+        HistoricalVersion.objects.filter(team_id=team_id, name=history_type, item_id=item_id).order_by("-versioned_at")[
+            :11
+        ]
+    )
+
+    if len(versions) == 0:
+        """
+        This item existed before history logging and this is the first time it's been viewed.
+        Create an import and capture the state as it is now
+        Otherwise the first change made by a user shows as a change to every field
+        """
+        imported_version = HistoricalVersion(
+            state=serializer(instance).data,
+            name="FeatureFlag",
+            action="update",
+            item_id=item_id,
+            created_by=_get_history_hog(),
+            team_id=team_id,
+        )
+        imported_version.save()
+
+        versions.append(imported_version)
```

Comment:
Given we have an unknown number of models this might be applied to 

(someone may only update to a version with a history log after we've added many types to track)

It seemed safer to add the import on demand than risk a migration that takes a long time